pkg/logql/syntax/ast.go

```diff
@@ -332,6 +332,19 @@ type LabelParserExpr struct {
 }
 
 func newLabelParserExpr(op, param string) *LabelParserExpr {
+	if op == OpParserTypeRegexp {
```

Comment:
Can we not handle this error higher up? A missing param to these parsers is a syntax error; specifically handling this here seems like we're addressing the symptom not the cause.

Reply:
I mean couldn't we handle this in the grammar?
It's strange to me that these queries are even submitted because they don't match the grammar

```yacc
...
  | REGEXP STRING  { $$ = newLabelParserExpr(OpParserTypeRegexp, $2) }
  | PATTERN STRING { $$ = newLabelParserExpr(OpParserTypePattern, $2) }
```

In any case, I'm surprised why this error isn't handled by the `func (m MultiStageExpr) stages()` function, because instantiating a new pattern parser with an invalid expression should return an error.

It feels awkward to special-case these here, and it makes it possible us to add new parsers later and forget to special-case those here too.

Reply:
These queries match the grammar. They pass an empty string ``. The STRING token in our grammar matches this. I'll see if I can create a special string type for these parsers that only matches if it's not an empty string.
The `Stage()` method only seems to be called from `Pipeline()`, which is done after parsing. Perhaps this was done to avoid creating a regex until needed.

Reply:
In the query you posted in the description, I don't see an empty string param:

```logql
topk(10,sum by(namespace)(count_over_time({application="nginx", site!="eu-west-1-dev"} |= "/artifactory/" != "api" != "binarystore" | regexp[1d])))
```

If no param is supplied, do you mean it defaults to the zero value of `STRING` which I guess would be `""`?

Reply:
That is actually not the query that was posted, it is the query printed when the original panic occurred. Susana found this in the [logs](https://admin-ops-us-east-0.grafana.net/grafana/explore?orgId=1&left=%7B%22datasource%22:%22loki-ops%22,%22queries%22:%5B%7B%22refId%22:%22A%22,%22datasource%22:%7B%22type%22:%22loki%22,%22uid%22:%22loki-ops%22%7D,%22editorMode%22:%22code%22,%22expr%22:%22%7Bcluster%3D%5C%22prod-us-central-0%5C%22,job%3D%5C%22loki-prod%2Fquery-frontend%5C%22%7D%20%21%3D%20%5C%22level%3D%5C%22%22,%22queryType%22:%22range%22,%22hide%22:true%7D,%7B%22refId%22:%22B%22,%22datasource%22:%7B%22type%22:%22loki%22,%22uid%22:%22loki-ops%22%7D,%22editorMode%22:%22code%22,%22expr%22:%22%7Bcluster%3D%5C%22prod-us-central-0%5C%22,job%3D%5C%22loki-prod%2Fquery-frontend%5C%22%7D%20%7C%3D%20%5C%22regexp%5B%5C%22%22,%22queryType%22:%22range%22%7D%5D,%22range%22:%7B%22from%22:%221670417578824%22,%22to%22:%221670421178824%22%7D%7D). Note the second line which has the GET command that has `regexp+%60%60`.  These last 2 characters translate to  ``.

Reply:
That test case can be removed if needed.

Reply:
I'd suggest updating the description to reflect the actual query that causes the panic though

Reply:
Done.